raidenchannels/raidenmanager.go

```diff
@@ -0,0 +1,165 @@
+package raidenchannels
+
+import (
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ExchangeUnion/xud-tests/ethclient"
+
+	"github.com/ExchangeUnion/xud-tests/xudclient"
+
+	"github.com/ExchangeUnion/xud-tests/raidenclient"
+	"github.com/ExchangeUnion/xud-tests/slackclient"
+)
+
+// TODO: clean inactive channels
+
+type token struct {
+	address       string
+	channelAmount uint64
+}
+
+var channelTokens = []token{
+	{
+		address:       "123",
+		channelAmount: 123,
+	},
+	{
+		address:       "987",
+		channelAmount: 987,
+	},
+}
+
+var raidenChannels = make(map[string]bool)
+
+// InitChannelManager initializes a new Raiden channel manager
+func InitChannelManager(
+	wg *sync.WaitGroup,
+	xud *xudclient.Xud,
+	raiden *raidenclient.Raiden,
+	eth *ethclient.Ethereum,
+	slack *slackclient.Slack) {
+
+	wg.Add(1)
+
+	initRaidenChannelsMap(raiden)
+
+	secondTicker := time.NewTicker(30 * time.Second)
+	dailyTicker := time.NewTicker(24 * time.Hour)
+
+	go func() {
+		defer wg.Done()
+
+		openChannels(xud, raiden, eth, slack)
+		balanceChannels(raiden, slack)
+
+		for {
+			select {
+			case <-secondTicker.C:
+				openChannels(xud, raiden, eth, slack)
+				balanceChannels(raiden, slack)
+				break
+
+			case <-dailyTicker.C:
+				raidenChannels = make(map[string]bool)
+				initRaidenChannelsMap(raiden)
+				break
+			}
+		}
+	}()
+}
+
+func initRaidenChannelsMap(raiden *raidenclient.Raiden) {
+	log.Debug("Querying and indexing existing Raiden channels")
+
+	// If the node has a channel for one token one can assume that the node has channels for the other tokens too
+	token := channelTokens[0]
+	channels, err := raiden.ListChannels(token.address)
+
+	if err != nil {
+		log.Error("Could not query channels of Raiden: %v", err.Error())
+		return
+	}
+
+	for _, channel := range channels {
+		raidenChannels[channel.PartnerAddress] = true
+	}
+}
+
+func openChannels(xud *xudclient.Xud, raiden *raidenclient.Raiden, eth *ethclient.Ethereum, slack *slackclient.Slack) {
+	log.Debug("Checking XUD for new Raiden partner addresses")
+
+	peers, err := xud.ListPeers()
+
+	if err != nil {
+		log.Error("Could not query XUD peers: %v", err.Error())
+		return
+	}
+
+	for _, peer := range peers.Peers {
+		_, hasChannels := raidenChannels[peer.RaidenAddress]
+
+		if !hasChannels {
+			raidenChannels[peer.RaidenAddress] = true
```

Comment:
Not sure it makes a difference but it might make sense to set this to `true` after the `openChannel` call completes below, or maybe better yet to set this to true within the `openChannel` function itself.

Reply:
Setting it to true after the call completes wouldn't work as Offer explained [here](https://github.com/ExchangeUnion/xud-simnet/issues/28#issuecomment-505728820):

> channels opening
for this, hold in-memory data structure of peers for which you already tried to open a channel. This data structure should prevent a loop of failed open channel and also prevent an overrun of two open channel.

The Raiden endpoint that opens a channel is blocking and will not return anything until the channel is actually open. If we set the value after the call is done, it could happen that the bot opens two or even more channels to each peer.

Reply:
> or maybe better yet to set this to true within the openChannel function itself.

It would come down to the exact same if this line is moved to the beginning of the `openChannel` function but you are right, it will be a little cleaner.

Reply:
Sounds good, thanks for the explanation.